rust/pattern/constraint/predicate.rs

```diff
@@ -86,88 +92,68 @@ impl fmt::Display for ValueConstraint {
 
 #[derive(Debug, Clone, PartialEq)]
 pub enum Value {
-    Long(i64),
-    Double(f64),
-    Boolean(bool),
-    String(String),
-    DateTime(NaiveDateTime),
-    Variable(Box<ThingVariable>),
+    Literal(Constant),
+    ThingVariable(Box<ThingVariable>),
+    ValueVariable(Box<ValueVariable>),
 }
 impl Eq for Value {} // can't derive, because floating point types do not implement Eq
 
 impl Validatable for Value {
     fn validate(&self) -> Result<()> {
         match &self {
-            Self::DateTime(date_time) => {
-                if date_time.nanosecond() % 1000000 > 0 {
-                    Err(TypeQLError::InvalidConstraintDatetimePrecision(*date_time))?
-                }
-                Ok(())
-            }
-            Self::Variable(variable) => variable.validate(),
-            _ => Ok(()),
+            Self::Literal(literal) => literal.validate(),
+            Self::ThingVariable(variable) => variable.validate(),
+            Self::ValueVariable(variable) => variable.validate(),
         }
     }
 }
 
-impl From<i64> for Value {
-    fn from(long: i64) -> Self {
-        Value::Long(long)
-    }
-}
-
-impl From<f64> for Value {
-    fn from(double: f64) -> Self {
-        Value::Double(double)
-    }
-}
-
-impl From<bool> for Value {
-    fn from(bool: bool) -> Self {
-        Value::Boolean(bool)
-    }
-}
-
-impl From<&str> for Value {
-    fn from(string: &str) -> Self {
-        Value::String(String::from(string))
+impl<T: Into<Constant>> From<T> for Value {
+    fn from(literal: T) -> Self {
```

Comment:
more `literal` -- do a search over the whole project for "literal" just in case :)